dvc/remote/gdrive/client.py

```diff
@@ -0,0 +1,284 @@
+    ):
+        # determine the offset
+        response = self.session.put(
+            resumable_upload_url,
+            headers={
+                "Content-Length": str(0),
+                "Content-Range": "bytes */{}".format(file_size),
+            },
+            timeout=self.TIMEOUT,
+        )
+        if response.status_code in (200, 201):
+            # file has been already uploaded
+            return True
+        elif response.status_code == 404:
+            # restarting upload from the beginning wouldn't make a
+            # profit, so it is better to notify the user
+            raise GDriveError("upload failed, try again")
+        elif response.status_code != 308:
+            logger.error(
+                "upload resume failure: {}".format(
+                    response_error_message(response)
+                )
+            )
+            return False
+        # ^^ response.status_code is 308 (Resume Incomplete) - continue
+        # the upload
+
+        if "Range" in response.headers:
+            # if Range header contains a string "bytes 0-9/20"
+            # then the server has received the bytes from 0 to 9
+            # (including the ends), so upload should be resumed from
+            # byte 10
+            offset = int(response.headers["Range"].split("-")[-1]) + 1
+        else:
+            # there could be no Range header in the server response,
+            # then upload should be resumed from start
+            offset = 0
+        logger.debug(
+            "resuming {} upload from offset {}".format(to_info, offset)
+        )
+
+        # resume the upload
+        from_file.seek(offset, 0)
+        response = self.session.put(
+            resumable_upload_url,
+            data=from_file,
+            headers={
+                "Content-Length": str(file_size - offset),
+                "Content-Range": "bytes {}-{}/{}".format(
+                    offset, file_size - 1, file_size
+                ),
+            },
+            timeout=self.TIMEOUT,
+        )
+        return response.status_code in (200, 201)
+
+    def upload(self, parent, to_info, from_file):
+        # Resumable upload protocol implementation
+        # https://developers.google.com/drive/api/v3/manage-uploads#resumable
+        resumable_upload_url = self._resumable_upload_initiate(
+            parent, posixpath.basename(to_info.path)
+        )
+        file_size = os.fstat(from_file.fileno()).st_size
+        success = self._resumable_upload_first_request(
+            resumable_upload_url, from_file, to_info, file_size
+        )
+        errors_count = 0
+        while not success:
+            try:
+                success = self._resumable_upload_resume(
+                    resumable_upload_url, from_file, to_info, file_size
+                )
+            except ConnectionError:
+                errors_count += 1
+                if errors_count >= 10:
+                    raise
+                sleep(1.0)
```

Comment:
do we actually need this sleep here?

Reply:
It is a must to wait some time between consecutive resumable upload requests. If one request would fail due to a short network problem then it would end up with all retry attempts failed in a short time. Maybe the same exponential backoff policy should be used here, as it is for error handling in `self.request`. Though it is not so clear for me would it be the right solution for resumable upload or not. The hardcoded 1 second sleep with 10 retries looks better, imho, but it is also not the best behavior, definitely.

One possible solution could be to store the upload process in the DVC's state database to make it possible to resume uploads between the dvc runs, but this feels like an overkill for me. Other backends don't care about the connection interruptions / server errors during large files upload at all, if I'm not mistaken.

Reply:
Yep. DB is an overkill for sure. As for retries - is it possible to use some decorator out of many existing? In both cases.